src/overlay/Peer.h

```diff
@@ -171,47 +165,84 @@ class Peer : public std::enable_shared_from_this<Peer>,
   protected:
     class MsgCapacityTracker : private NonMovableOrCopyable
     {
-        std::weak_ptr<Peer> mWeakPeer;
-        StellarMessage mMsg;
+        std::weak_ptr<Peer> const mWeakPeer;
+        StellarMessage const mMsg;
+        std::atomic<bool> mFinished{false};
 
       public:
         MsgCapacityTracker(std::weak_ptr<Peer> peer, StellarMessage const& msg);
-        ~MsgCapacityTracker();
         StellarMessage const& getMessage();
-        std::weak_ptr<Peer> getPeer();
+        // Since MsgCapacityTracker is passed around between thread and stored
+        // by ASIO in lambdas, do not rely on RAII to trigger completion.
+        // Instead, explicitly call `finish`.
+        void finish();
     };
 
     OverlayAppConnector mAppConnector;
 
     Hash const mNetworkID;
     std::shared_ptr<FlowControl> mFlowControl;
-    VirtualClock::time_point mLastRead;
-    VirtualClock::time_point mLastWrite;
-    VirtualClock::time_point mEnqueueTimeOfLastWrite;
+    std::atomic<VirtualClock::time_point> mLastRead;
+    std::atomic<VirtualClock::time_point> mLastWrite;
+    std::atomic<VirtualClock::time_point> mEnqueueTimeOfLastWrite;
 
     PeerRole const mRole;
     OverlayMetrics& mOverlayMetrics;
     PeerMetrics mPeerMetrics;
 
+    // Mutex to protect PeerState, which can be accessed and modified from
+    // multiple threads
+#ifndef USE_TRACY
+    std::recursive_mutex mutable mStateMutex;
+#else
+    mutable TracyLockable(std::recursive_mutex, mStateMutex);
+#endif
+
     HmacSha256Key mSendMacKey;
     HmacSha256Key mRecvMacKey;
-    uint64_t mSendMacSeq{0};
-    uint64_t mRecvMacSeq{0};
-    VirtualTimer mRecurringTimer;
-
+    std::atomic<uint64_t> mSendMacSeq{0};
+    std::atomic<uint64_t> mRecvMacSeq{0};
     // Does local node have capacity to read from this peer
     bool canRead() const;
     // helper method to acknowledge that some bytes were received
     void receivedBytes(size_t byteCount, bool gotFullMessage);
+    virtual bool
+    useBackgroundThread() const
+    {
+        return mAppConnector.getConfig()
+            .EXPERIMENTAL_BACKGROUND_OVERLAY_PROCESSING;
+    }
+
     void initialize(PeerBareAddress const& address);
-    void setState(PeerState newState);
-    bool shouldAbort() const;
+    void setState(LockGuard const& stateGuard, PeerState newState);
+    bool shouldAbort(LockGuard const& stateGuard) const;
+    void shutdownAndRemovePeer(std::string const& reason,
+                               DropDirection dropDirection);
+
+    // Subclasses should only use these methods to load and modify peer state,
+    // signatured hint that callers need to grab mStateMutex.
+    PeerState
+    getState(LockGuard const& stateGuard) const
+    {
+        return mState;
+    }
 
```

Comment:
Small suggestion here to reorganize the signatures and comment. Also it seems to me odd that 2/3 of the methods are implemented out-of-line -- maybe make the 3rd also out-of-line too, for symmetry? It's not like it's a performance bottleneck.

```suggestion
    void shutdownAndRemovePeer(std::string const& reason,
                               DropDirection dropDirection);

    // Subclasses should only use these methods to access peer state;
    // they all take a LockGuard that should be holding mStateMutex,
    // but do not lock that mutex themselves (to allow atomic
    // read-modify-write cycles or similar patterns in callers).

    bool shouldAbort(LockGuard const& stateGuard) const;
    void setState(LockGuard const& stateGuard, PeerState newState);
    PeerState
    getState(LockGuard const& stateGuard) const
    {
        return mState;
    }

```